Slim/Utils/Prefs.pm

```diff
@@ -394,9 +394,30 @@ sub init {
 
 	$prefs->setChange(
 		sub { Slim::Control::Request::executeRequest(undef, ['wipecache', $prefs->get('dontTriggerScanOnPrefChange') ? 'queue' : undef]) },
-		qw(splitList groupdiscs useTPE2AsAlbumArtist userDefinedRoles)
+		qw(splitList groupdiscs useTPE2AsAlbumArtist)
 	);
 
+	$prefs->setChange( sub {
+		my $newRoles = $_[1];
+		my $oldRoles = $_[3];
+		my $changed = 0;
+		if (%$oldRoles != %$newRoles) {
```

Comment:
This comparison would not compare the hash's content, would it? You might have to compare keys by key (or values) yourself anyway. Check number of items first. That's cheap and covers additions/removals. Then compare key by key. No need for a copy of one of the hashes. Just look them up, and upon first failure stop and set the changed flag.

Reply:
As far as I understand it (google!) %$oldRoles != %$newRoles is a scalar comparison of the number of items. If different, the code goes on to check each key old vs new.

I got the code from here: https://stackoverflow.com/a/1273816

Reply:
Ha! You taught me a lesson today 😁. Thanks!

Reply:
But I still don't get why you'd need `%cmp`. Why not compare to the original hash directly?

Reply:
I think because it's deleting from %cmp when the new key matches - an empty %cmp at the end means nothing changed.

Yes, we could retrieve the pref for $oldRoles, but as it's passed in @_ anyway...

Reply:
But we don't have to delete, do we? After we've figured out they're the same size we compare all keys/values from one hash with the other. This isn't destructive, is it?

Reply:
So far I just copied the code from the stackoverflow post. What's the best way of doing the compare?

Reply:
I'm no expert for "best" solutions 😁. But here's what I think should do the job:

```perl
while (my ($k, $v) = each %oldRoles) {
   if (!exists $newRoles{$k} || $newRoles{$k} ne $v) {
      $changed = 1;
      last;
   }
}
```

Your change only checks for the existence of a key, but not for its content.

Reply:
Yes, that is intentional, the key is the tag name to be scanned, the values are the display text and the "include" flag which should not trigger a rescan. It's the whole reason this new `setChange` sub is necessary.

Reply:
Ohh... ok. In that case you can obviously remove the second condition in my statement.

I think `$changed` and `%cmp` server the same purpose: have an indicator whether something has changed. If we didn't remove everything from `%cmp` then we'd trigger a rescan. But it's kind of redundant in this case: we would set `$changed = 1` if the temporary hash copy wasn't empty. But we can as well just set the flag when we evaluate the existence of the key in both hashes. That would get rid of the `if (%cmp)` check.

Just for fun here's probably another, more perly way of achieving the same:
```perl
$changed = %oldRoles - %newRoles || (scalar grep {!exists $newRoles{$_}} keys %oldRoles);
```

The subtraction would check the element number, the second part would check whether all the keys were the same...

Reply:
I like perly!

See latest push, which hopefully resolves all issues.

Do you want to merge #1175 so that I can merge public/9.0 into this one before doing a final push?